Zendesk: support user account subdomain configuration

spolu opened the pull request:
## Description

Fixes: https://github.com/dust-tt/tasks/issues/1517

- adds support for extra_config being passed through the setup redirect URL as metadata on the Connection
- Tweak Zendesk setup screen to capture the Zendesk account subdomain
- Change oauth zendesk adapter to pull the subdomain from the metadata of the Connection (and check it)
- Move zendesk to being rolling_out (feature flag based)

![Screenshot from 2024-10-31 15-59-05](https://github.com/user-attachments/assets/27e8d646-eb8f-4a87-bb79-c9ecc9a014a1)

## Risk

Low, not released yet. Did a successful setup in dev.

## Deploy Plan

- deploy `oauth
- deploy `front`

r? @aubin-tchoi cc @albandum

spolu commented on front/lib/connector_providers.ts:
```diff
@@ -187,7 +187,7 @@ export const CONNECTOR_CONFIGURATIONS: Record<
     name: "Zendesk",
     connectorProvider: "zendesk",
-    status: "preview",
+    status: "rolling_out",
```
@aubin-tchoi hope that's ok?

spolu commented:
Given that it touches oauth will ask for another reviewer r? @flvndvd

flvndvd commented on core/src/oauth/providers/zendesk.rs:
```diff
@@ -45,8 +48,20 @@ impl Provider for ZendeskConnectionProvider {
             "scope": "read"
         });
 
+        let sub_domain = match connection.metadata()["extra_config"].as_str() {
```
Shouldn't we define `extra_config` as an object to support more in the future? Which seems very likely something we will need.

spolu replied:
Well technically metadata is a Value in `oauth` so we definitely could, but at the same time we won't get much guarantee on the typing through `oauth` so that would be a pure `front`/`connectors` convention.

spolu replied:
The whole point of metadata is to be that very object so I'm bearish of adding an object in an object. I can refactor the code to return an extraConfig Record<string, string> at setup that gets flattened in the metadata so that core will store `zendesk_subdomain` in metadata instead of an opaque `extra_config`. I think that's the best approach, wdyt?

flvndvd replied:
Yes as long as it's clearer what `extra_config` holds I'm fine.

spolu replied:
extra_config will just live on front during the connection setup. The problem is that whatever we build we will have to pass as GET parameter.... I presume we can just stringify (maybe base64 encode) it.

flvndvd commented on front/components/data_source/CreateConnectionConfirmationModal.tsx:
```diff
@@ -24,6 +25,15 @@ export function CreateConnectionConfirmationModal({
   const [isLoading, setIsLoading] = useState(false);
+  const [extraConfig, setExtraConfig] = useState<string | null>(null);
+
+  const isExtraConfigValid = useCallback(() => {
+    if (connectorProviderConfiguration.connectorProvider === "zendesk") {
+      return /^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$/.test(extraConfig || "");
+    } else {
+      return true;
+    }
+  }, [connectorProviderConfiguration.connectorProvider, extraConfig]);
```
Not ideal, as this function will be re-callbackify on each key stroke in `extraConfig`. Instead, you should pass `extraConfig` as a function param.

spolu replied:
sure thing

flvndvd commented on core/src/oauth/providers/zendesk.rs:
```diff
@@ -45,8 +48,20 @@ impl Provider for ZendeskConnectionProvider {
+
         let req = reqwest::Client::new()
-            .post("https://d3v-dust.zendesk.com/oauth/tokens")
+            .post(url)
```
Any chance we can validate the subdomain in the response as well?

spolu replied:
wdym in the response? we're hitting it, there nothing to validate no?

flvndvd replied:
Was mainly wondering if we need to add some specific checks in the response handling.

spolu replied:
I don't believe we need to add anything here.

aubin-tchoi commented on front/lib/api/oauth.ts:
```diff
@@ -247,8 +247,12 @@ const PROVIDER_STRATEGIES: Record<
   zendesk: {
     setupUri: (connection) => {
       const scopes = ["read"];
+      let subdomain = "not-valid-will-fail";
```
does this fail tho?

spolu replied:
(updated)

spolu commented:
Updated to handle an extraConfig object as we setup the connection which then gets flattened in the metadata. It's flexible though not strictly typed and has the advantage that the oauth record is clean and ends up with `zendesk_subdomain` in its metadata which will be usable more explicitely from connectors.

I also added strict validation of the extraConfig we receive as GET parameter before creating the connection to prevent anyone from spamming the content of our OAUth DB.

flvndvd commented on front/components/data_source/CreateConnectionConfirmationModal.tsx:
```diff
@@ -24,10 +26,21 @@ export function CreateConnectionConfirmationModal({
+    } else {
+      return true;
+    }
+  }, [connectorProviderConfiguration.connectorProvider, extraConfig]);
```
Ditto the previous comment, let's move `extraConfig` as a function parameters so it does not get re-callbackify on every key stroke.

flvndvd commented on front/pages/w/[wId]/oauth/[provider]/setup.tsx:
```diff
@@ -11,24 +16,41 @@ export const getServerSideProps = withDefaultUserAuthRequirements<object>(
+    let parsedExtraConfig: Record<string, string> = {};
+    try {
+      const bodyValidation = ExtraConfigTypeSchema.decode(
+        JSON.parse(extraConfig as string)
```
We have `safeParse` helper so you can remove the try/catch.

flvndvd reviewed (approved):
LGTM once the two last comments have been fixed.

spolu commented:
Connection setup has been thoroughly tested in dev. Deploying and will test a random non zendesk connection setup in production
